### ISSUE TITLE
Support for long sequences classification (transformer models)

### ISSUE THREAD

**Issue by mdavis95** [amaiya/ktrain#84]
It would be great if support could be added for long sequences over ~500 word pieces for the transformers models.

Possible methods:
1)  sliding window over each sequence to generate sub-sequences output that is averaged before being feed to the classifier layer
2) sliding window over each sequence to generate sub-sequences output feed to LSTM layer

**Comment by amaiya** [comment]
Thanks - good suggestion.

**Comment by basque21** [comment]
hi, I'm also interested on that. What is the maximum length at this moment? 
Thanks

**Comment by mdavis95** [comment]
Depends on the model I believe but generally 512

**Comment by basque21** [comment]
thanks @mdavis95 !

**Comment by Niekvdplas** [comment]
Hi @mdavis95,

I hope I get what you are saying and I wrote a little script which does exactly this. Splitting on sequences of around 500 word pieces. I tried to first find one of the last lines/sentences and if that didnt work just select first 512. Here is my code, maybe it might help you.

`
#texts is the full text read in.
textlist = []
        startpos = 0
        while len(texts[startpos:]) > 512:
            idx = texts[startpos:startpos + 512].rfind('\n')
            if idx == -1:
                idx = texts[startpos:startpos + 512].rfind('.')
            if idx == -1:
                idx = texts[startpos:startpos + 512].rfind(' ')
            if idx == -1:
                idx = 512
            textlist.append(texts[startpos:startpos + idx])
            startpos = startpos + idx + 1
        textlist.append(texts[startpos:])

        return textlist
`

**Comment by marcturpin91** [comment]
@Niekvdplas I believe what @mdavis95 wants is a way to not only separate text like you did by chunks, but to be able to feed a ktrain model multiple chunks that belong to a single sample. For example, if we're classifying a document that is 2000 words, we can divide it into 4 chunks of 500 words. The issue is we will get 4 outputs when in reality we are looking to classify one document that has been separated into 4 pieces.

We can manually take those outputs and do some type of calculations (average, max, etc..) to get the final label, but the model is not learning from this final step given we're doing it after the classifying step.

**Comment by mdavis95** [comment]
Sorry I missed the original comment.  The summary of what I was looking for was perfect.

**Comment by marcturpin91** [comment]
@amaiya is there any way to make the tokenizer return the x first tokens + the last y tokens? This research paper [https://arxiv.org/abs/1905.05583](url) "How to Fine-Tune BERT for Text Classification?" mentions that when processing long text sequences that exceeded the 512 token limit, they found the best results by taking the first 128 tokens and the last 382. 

If not, do you have any suggestions as to how I could achieve this while using ktrain?

**Comment by amaiya** [comment]
Hi @marcturpin91  

The truncation happens within the methods of the tokenizer (e.g,. BertTokenizer) such as `encode_plus`.  The `transformer` libarary tokenization methods have a `truncation_strategy` option, but it's not designed to support what you described.   So, unless you modify the transformer library itself by adding additional truncation strategies, the way to do it is probably preprocess the data prior to it being tokenized.  This could be done either inside or outside of *ktrain*, but it would probably make sense to test it outside of *ktrain* for now.  The basic idea is to:
1. tokenize the text, which will convert some words into subwords.  e.g.,  "embeddings" becomes `['em', '##bed', '##ding', '##s']`.  
2. Truncate using the strategy you described (i.e., take first x tokens and last y tokens). 
3. [optional]  re-construct subwords into words again.  `['em', '##bed', '##ding', '##s']` --> 'embeddings'
4. Feed this as input to `preprocess_train` or `texts_from_array`

Steps 1 and 3 are easy to do, as the BertTokenizer already has methods for doing this.   You can refer to the `transformer` tokenizer documentation and source code for more details.

**Comment by mdavis95** [comment]
looks like Longformer is supported by HuggingFace now (https://huggingface.co/transformers/model_doc/longformer.html).  It seems to support 4k tokens with better performance.  Still not the exact answer but does help in practice

**Comment by xirect** [comment]
I couldn't get the Longformer models working as KTrain requires <4.0.0 transformers version.

**Comment by amaiya** [comment]
#### Longformer

```python
# load text data
categories = ['alt.atheism', 'soc.religion.christian','comp.graphics', 'sci.med']
from sklearn.datasets import fetch_20newsgroups
train_b = fetch_20newsgroups(subset='train', categories=categories, shuffle=True)
test_b = fetch_20newsgroups(subset='test',categories=categories, shuffle=True)
(x_train, y_train) = (train_b.data, train_b.target)
(x_test, y_test) = (test_b.data, test_b.target)

# build and train Longformer model
import ktrain
from ktrain import text
MODEL_NAME = 'allenai/longformer-base-4096'
t = text.Transformer(MODEL_NAME, maxlen=1024, class_names=train_b.target_names)
trn = t.preprocess_train(x_train, y_train)
model = t.get_classifier()
learner = ktrain.get_learner(model, train_data=trn, batch_size=1)
learner.fit_onecycle(5e-5, 1)
```
